shell/platform/darwin/macos/framework/Source/FlutterEngine.mm

```diff
@@ -411,29 +410,28 @@ - (void)loadAOTData:(NSString*)assetsDir {
 }
 
 - (void)setViewController:(FlutterViewController*)controller {
-  if (_viewController != controller) {
+  if (_viewController == controller) {
+    // From nil to nil, or from non-nil to the same controller.
+    return;
+  }
+  if (_viewController == nil && controller != nil) {
     _viewController = controller;
-
-    if (_semanticsEnabled && _bridge) {
-      _bridge->UpdateDefaultViewController(_viewController);
-    }
-
-    if (!controller && !_allowHeadlessExecution) {
+    [_viewController attachToEngine:self withId:kFlutterDefaultViewId];
+  } else if (_viewController != nil && controller == nil) {
+    [_viewController detachFromEngine];
+    _viewController = nil;
+    if (!_allowHeadlessExecution) {
       [self shutDownEngine];
     }
+  } else {
+    NSLog(@"Failed to set view controller to the engine: "
+          @"Replacing the view controller of the engine is not supported.");
```

Comment:
Should this crash in debug mode?

Reply:
I think assertions are for internal errors, and "misuses" should trigger something else, either an exception (which is discouraged in C++ and ObjC), an error code, or a log. But I can change it if you still support so.

Reply:
I'm worried a developer might try to replace a flutter view controller without realizing their code doesn't actually work. I don't think the NSLog is a strong enough signal.

FYI it looks like there are `NSAssert`s to validate misuses, for example: https://github.com/flutter/engine/blob/78bbea005d27e77193793cda562d5d8ef53070b4/shell/platform/darwin/ios/framework/Source/FlutterView.mm#L23-L26

Reply:
Done. I've also added fixing instructions to the assertions.